Implement steam features Part 1

dareumnam opened the pull request:
Pull request overview
---------------------
 - Enable `LoadProfile:Plant` to be used in a steam loop (`PlantLoadProfile.cc`)
 - Add `DistrictHeatingSteam` (`OutsideEnergySources.cc`)
 - Change the resource type name, the current object name and meter name of `DistrictHeating` to `DistrictHeatingWater`, and `Steam` to `DistrictHeatingSteam`

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [X] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

dareumnam committed:
PlantLoadProfileSteam

dareumnam committed:
PlantLoadProfileSteam add idd

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
Update PlantLoadProfile Steam

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
Add a PlantLoadProfile testfile

dareumnam committed:
clang format

dareumnam committed:
DistrictHeatingSteam plant setup

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
districtheatingsteam partial

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
OutsideEnergySources partial

dareumnam committed:
OutsideEnergySources partial

dareumnam committed:
add comments

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
DistrictHeating Steam Add OutputReoprts

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
DistrictHeatingSteam as resource

dareumnam committed:
DistrictHeatingSteam as resource

dareumnam committed:
DistrictHeatingSteam in EconomicTariff

dareumnam committed:
OutputReportTabular cleanup

dareumnam committed:
OutputReportTabular cleanup

dareumnam committed:
add a testfile

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
github token test

dareumnam committed:
SteamToWaterHX Plant setup

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
fix PlantLoadProfile ProfileNum Counter & Implement SteamToWaterHX 1

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
SteamToWaterHX:Typenum

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
SteamToWaterHX Outputvars

dareumnam committed:
Update SteamToWaterHX Outputvars

dareumnam committed:
Fix districtheatingsteam

dareumnam committed:
Update testing files for plantloadprofilesteam and districtheatingsteam

dareumnam committed:
Update idd and getinput() for steamtowaterHX

dareumnam committed:
Update Getinput & Idd & Initialize

dareumnam committed:
Pull develop and Merge conflicts

dareumnam committed:
SteamtoHX Calculate

dareumnam committed:
fix PLP Steam SetupOutputVariable arguments

dareumnam committed:
SteamtoHX Calculate 2

dareumnam committed:
Sizing

dareumnam committed:
Merge remote-tracking branch 'upstream/develop' into SteamToWaterHX

dareumnam committed:
Merge branch 'SteamToWaterHX' of https://github.com/dareumnam/EnergyPlus into SteamToWaterHX

dareumnam committed:
Update functions

dareumnam committed:
Add a new testfile for STWHX

dareumnam committed:
fix warnings

dareumnam committed:
Fix functions and unit tests

dareumnam committed:
update module

dareumnam committed:
fix ctrlftn

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
fix custom check errors in testfiles

dareumnam committed:
clang format

dareumnam committed:
DistrictHeatingSteam: Fix merge conflicts in OutputReportTabular.cc

dareumnam committed:
STWHX: Update testfile

dareumnam committed:
pull develop and merge conflict

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
update testing file

dareumnam committed:
cleanup

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModules

dareumnam committed:
cleanup

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModules

dareumnam committed:
pull develop and merge lots of conflicts

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
merge conflict

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
Merge conflicts

dareumnam committed:
cleanup

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
Edit idd

dareumnam committed:
Edit GetInputFunction

dareumnam committed:
update functions 1

dareumnam committed:
update functions 2

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
cleanup and add a testfile

dareumnam committed:
update design document 1

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
update testfiles cMakelist

dareumnam committed:
remove unit test temporarily

dareumnam committed:
Combine LoadProfile:Plant and LoadProfile:Plant:Steam

dareumnam committed:
Combine LoadProfile:Plant and LoadProfile:Plant:Steam 2

dareumnam committed:
Combine LoadProfile:Plant and LoadProfile:Plant:Steam 3

dareumnam committed:
add Num at enum class

dareumnam committed:
Edit testfiles

dareumnam committed:
add a Loadprofile:Plant input to testfiles

dareumnam committed:
Fix districtheatingsteam

dareumnam committed:
Fix districtheatingsteam 2

dareumnam committed:
DistrictHeating:Water and DistrictHeating:Steam 1

dareumnam committed:
DistrictHeating:Water and DistrictHeating:Steam 2

amirroth commented on src/EnergyPlus/DataGlobalConstants.cc:
```diff
@@ -129,6 +129,9 @@ ResourceType AssignResourceTypeNum(std::string const &ResourceTypeChar)
         } else if (SELECT_CASE_var == "DISTRICTHEATING") {
```
Use `getEnumerationValue` instead of a custom function.

dareumnam replied:
I'll eliminate these functions and use `getEnueratationVaule` and `std::array<std::string_view>`s on a separate branch for enum.

amirroth commented on src/EnergyPlus/DataGlobalConstants.cc:
```diff
@@ -286,6 +289,9 @@ std::string GetResourceTypeChar(ResourceType const ResourceTypeNum)
     case ResourceType::DistrictHeating: {
         return "DistrictHeating";
     } break;
+    case ResourceType::DistrictHeatingSteam: {
```
Use `constexpr std::array<std::string_view>` instead of a function.

amirroth commented on src/EnergyPlus/DataGlobalConstants.hh:
```diff
@@ -129,6 +129,7 @@ namespace DataGlobalConstants {
         Condensate,
         OtherFuel1,
         OtherFuel2,
+        DistrictHeatingSteam,
```
Does numbering make a difference here? Should this be moved next to `DistrictHeating` and `DistrictCooling`? 📌  need to do something about this and FuelType. Not in this branch.

amirroth commented on src/EnergyPlus/DataGlobalConstants.hh:
```diff
@@ -256,7 +257,8 @@ struct DataGlobalConstantsData : BaseGlobalStruct
                                                                        DataGlobalConstants::ResourceType::Condensate,
                                                                        DataGlobalConstants::ResourceType::OtherFuel1,
-                                                                       DataGlobalConstants::ResourceType::OtherFuel2};
+                                                                       DataGlobalConstants::ResourceType::OtherFuel2,
```
This data structure is nuts. The one right above it is worse.

amirroth commented on src/EnergyPlus/DataLoopNode.hh:
```diff
@@ -515,6 +517,7 @@ namespace DataLoopNode {
         "Dehumidifier:Desiccant:System",
```
These `constexpr std::array<std::string_view>` arrays should be moved to the .cc` files. If they are needed outside of this module you can leave an `extern const` declaration here. [Sorry, this is a new thing.]

amirroth commented on src/EnergyPlus/ExteriorEnergyUse.cc:
```diff
@@ -596,6 +596,9 @@ namespace ExteriorEnergyUse {
         } else if (UtilityRoutines::SameString(FuelTypeAlpha, "DistrictHeating")) {
             FuelTypeNumber = ExteriorEnergyUse::ExteriorFuelUsage::DistrictHeatUse;
             FuelTypeString = "DistrictHeating";
+        } else if (UtilityRoutines::SameString(FuelTypeAlpha, "DistrictHeatingSteam")) {
```
`getEnumerationValue`

amirroth commented on src/EnergyPlus/ExteriorEnergyUse.hh:
```diff
@@ -82,6 +82,7 @@ namespace ExteriorEnergyUse {
         DistrictHeatUse,
         OtherFuel1Use,
         OtherFuel2Use,
+        DistrictHeatSteamUse,
```
Man, really really need to unify these separate `FuelType` definitions.

amirroth commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -1693,6 +1693,9 @@ namespace OutputProcessor {
             } else if (meterType == "DISTRICTHEATING") {
                 OutResourceType = "DistrictHeating";
 
+            } else if (meterType == "DISTRICTHEATINGSTEAM") {
```
Using an `if-else` ladder to map strings to other strings. Yikes!

amirroth commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -8117,11 +8120,12 @@ void InitPollutionMeterReporting(EnergyPlusData &state, std::string const &Repor
     // SUBROUTINE PARAMETER DEFINITIONS:
     //             Now for the Pollution Meters
-    static Array1D_string const PollutionMeters({1, 29},
+    static Array1D_string const PollutionMeters({1, 30},
```
Is 30 actually `ResourceType::Num` or `SomethingElse::Num`?

amirroth replied:
Also, this should be a `constexpr std::array<std::string_view>`, not `Array1D_string`.

amirroth commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -1747,6 +1747,7 @@ void GetInputOutputTableSummaryReports(EnergyPlusData &state)
         ort->resourceTypeNames(12) = "Propane";
         ort->resourceTypeNames(13) = "OtherFuel1";
         ort->resourceTypeNames(14) = "OtherFuel2";
+        ort->resourceTypeNames(15) = "DistrictHeatingSteam";
```
Mmmmm, there's got to be a better way to do this. 📌  (this means "for later")

amirroth commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -7812,13 +7830,13 @@ void WriteBEPSTable(EnergyPlusData &state)
-    Array3D<Real64> collapsedEndUseSub(state.dataOutputProcessor->MaxNumSubcategories, DataGlobalConstantsData::iEndUseSize, 13);
-    Array2D<Real64> endUseSubOther(13, DataGlobalConstantsData::iEndUseSize);
-    Array3D<Real64> collapsedEndUseSpType(state.dataOutputProcessor->maxNumEndUseSpaceTypes, DataGlobalConstantsData::iEndUseSize, 13);
+    Array2D<Real64> useVal(14, 15);
```
I can't even. :pushpin:

amirroth commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -8018,7 +8041,7 @@ void WriteBEPSTable(EnergyPlusData &state)
 
             // convert units into GJ (divide by 1,000,000,000) if J otherwise kWh
-            for (iResource = 1; iResource <= 12; ++iResource) { // don't do water
+            for (iResource = 1; iResource <= 13; ++iResource) { // don't do water
```
I'm going to skip the rest of this file, because this could take hours.

amirroth commented on src/EnergyPlus/OutsideEnergySources.cc:
```diff
@@ -331,23 +367,55 @@ void OutsideEnergySourceSpecs::size(EnergyPlusData &state)
     if (this->EnergyType == DataPlant::PlantEquipmentType::PurchChilledWater) {
         typeName = "Cooling";
-    } else { // Heating
+    } else if (this->EnergyType == DataPlant::PlantEquipmentType::PurchHotWater) {
```
This `if-else` could be a lookup in a `constexpr std::array<std::string_view>`, no?

amirroth replied:
Also, just create the concatenated `"SizeDistrict" + typeName` string once here, and then reuse.

mjwitte replied:
As @amirroth suggested long ago, `typename` should be the actual object type, e.g. "DistrictHeating:Water" etc.
    `std::string_view typeName = DataPlant::PlantEquipTypeNames[static_cast<int>(this->EnergyType)];`

This replaces the entire if/elseif block. And everywhere that `typeName` is used, remove "District" from the format strings. @Myoldmopar I can make these changes and push if you want.

This should fix the eio component sizing outputs which are currently like:
```
Component Sizing Information, DistrictHeatingWater, PURCHASED HEATING, User-Specified Nominal Capacity [W], 1000000.00000
```
to be
```
Component Sizing Information, DistrictHeating:Water, PURCHASED HEATING, User-Specified Nominal Capacity [W], 1000000.00000
```

Myoldmopar replied:
@mjwitte that would be fine.  As of now I don't have any changes to the C++ code, just transition changes.  So if you change this it wouldn't conflict and would only clean up my diffs.

mjwitte replied:
Done, but this won't reduce the diffs, it'll just make them correct.

Myoldmopar replied:
Ahhh that's fine anyway.  Thanks

amirroth commented on src/EnergyPlus/OutsideEnergySources.cc:
```diff
@@ -331,23 +367,55 @@ void OutsideEnergySourceSpecs::size(EnergyPlusData &state)
+    } else { // DataPlant::PlantEquipmentType::PurchSteam
+        typeName = "HeatingSteam";
     }
 
```
Can make an `auto &loop = state.dataPlnt->PlantLoop(this->plantLoc.loopNum);` shortcut and use it throughout here.

amirroth commented on src/EnergyPlus/OutsideEnergySources.cc:
```diff
@@ -331,23 +367,55 @@ void OutsideEnergySourceSpecs::size(EnergyPlusData &state)
                                                                  state.dataPlnt->PlantLoop(this->plantLoc.loopNum).FluidName,
                                                                  DataGlobalConstants::InitConvTemp,
                                                                  state.dataPlnt->PlantLoop(this->plantLoc.loopNum).FluidIndex,
                                                                  "SizeDistrict" + typeName);
```
Use `format` instead of string concatenation, especially for constant strings.

amirroth commented on src/EnergyPlus/OutsideEnergySources.cc:
```diff
@@ -427,19 +496,30 @@ void OutsideEnergySourceSpecs::calculate(EnergyPlusData &state, bool runFlag, Re
         if (MyLoad > 0.0) MyLoad = 0.0;
     } else if (this->EnergyType == DataPlant::PlantEquipmentType::PurchHotWater) {
         if (MyLoad < 0.0) MyLoad = 0.0;
+    } else if (this->EnergyType == DataPlant::PlantEquipmentType::PurchSteam) {
```
This is the same case as the previous one, basically. Can use `||`.

amirroth commented on src/EnergyPlus/OutsideEnergySources.cc:
```diff
@@ -469,18 +549,24 @@ void OutsideEnergySourceSpecs::oneTimeInit_new(EnergyPlusData &state)
+
     SetupOutputVariable(state,
-                        reportVarPrefix + hotOrChilled + "Water Energy",
+                        reportVarPrefix + hotOrChilled + "Energy",
```
`format`

amirroth commented on src/EnergyPlus/Plant/DataPlant.hh:
```diff
@@ -182,7 +182,9 @@ namespace DataPlant {
         "AirTerminal:SingleDuct:ConstantVolume:FourPipeBeam",
         "ZoneHVAC:CoolingPanel:RadiantConvective:Water",
         "HeatPump:PlantLoop:EIR:Cooling",
-        "HeatPump:PlantLoop:EIR:Heating"};
```
Move this `std::array` to the `.cc` file.

amirroth commented on src/EnergyPlus/Plant/DataPlant.hh:
```diff
@@ -182,7 +182,9 @@ namespace DataPlant {
+        "DistrictHeatingSteam",
+        "HeatExchanger:SteamToWater"};
 
     static constexpr std::array<std::string_view, static_cast<size_t>(PlantEquipmentType::Num)> PlantEquipTypeNamesUC{
```
And this one.

amirroth commented on src/EnergyPlus/Plant/DataPlant.hh:
```diff
@@ -280,7 +282,9 @@ namespace DataPlant {
+        "DISTRICTHEATINGSTEAM",
+        "HEATEXCHANGER:STEAMTOWATER"};
 
     static constexpr std::array<LoopType, static_cast<size_t>(PlantEquipmentType::Num)> ValidLoopEquipTypes{
```
And this one.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
+        } else {
+            state.dataPlantHXFluidToFluid->FluidHX(CompLoop).AvailSchedNum = ScheduleManager::GetScheduleIndex(state, cAlphaArgs(2));
+            if (state.dataPlantHXFluidToFluid->FluidHX(CompLoop).AvailSchedNum <= 0) {
+                ShowSevereError(state, std::string{RoutineName} + cCurrentModuleObject + "=\"" + cAlphaArgs(1) + "\", invalid entry.");
```
`format` instead of `+`.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
+                                                                 cAlphaFieldNames,
+                                                                 cNumericFieldNames);
+        UtilityRoutines::IsNameEmpty(state, cAlphaArgs(1), cCurrentModuleObject, ErrorsFound);
+
```
An `auto &thisFluidHX = state.dataPlantHXFluidToFluid->FluidHX(CompLoop);` shortcut would be handy here.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
+            state.dataPlantHXFluidToFluid->FluidHX(CompLoop).SupplySideLoop.DesignVolumeFlowRateWasAutoSized = true;
+        }
+
+        if (UtilityRoutines::SameString(cAlphaArgs(7), "CrossFlowBothUnMixed")) {
```
`getEnumerationValue`.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
+        }
 
-            state.dataPlantHXFluidToFluid->FluidHX(CompLoop).DemandSideLoop.inletNodeNum =
+        if (UtilityRoutines::SameString(cAlphaArgs(8), "UncontrolledOn")) {
```
`getEnumerationValue`

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
                                                     DataLoopNode::ObjectIsNotParent);
-            state.dataPlantHXFluidToFluid->FluidHX(CompLoop).DemandSideLoop.outletNodeNum =
+            // check that node actually has setpoints on it
+            if ((state.dataPlantHXFluidToFluid->FluidHX(CompLoop).controlMode == ControlType::HeatingSetPointModulated) ||
```
`switch/case` or `BITF`.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
-                    ShowContinueError(state, "Missing entry for " + cNumericFieldNames(3));
-                    ErrorsFound = true;
-                }
+        if (!lAlphaFieldBlanks(13)) {
```
`getEnumerationValue`.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
-            } else if (UtilityRoutines::SameString(cAlphaArgs(8), "TrackComponentOnOff")) {
-                state.dataPlantHXFluidToFluid->FluidHX(CompLoop).controlMode = ControlType::TrackComponentOnOff;
-            } else {
+        if (!lNumericFieldBlanks(5)) {
```
A shorter idiom here would be `thisFluidHX.SizingFactor = !lNumericFieldBlanks(5) ? rNumericArgs(5) : 1.0;`

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
+            ShowContinueError(state, "Missing entry for " + cNumericFieldNames(3));
+            ErrorsFound = true;
+        }
+
```
You know what I am going to say here.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
+                    (state.dataLoopNodes->Node(state.dataPlantHXFluidToFluid->FluidHX(CompLoop).SetPointNodeNum).TempSetPointLo ==
+                     DataLoopNode::SensedNodeFlagValue)) {
+                    if (!state.dataGlobal->AnyEnergyManagementSystemInModel) {
+                        ShowSevereError(state, std::string{RoutineName} + " Missing dual temperature setpoints for node = " + cAlphaArgs(8));
```
You know what I am going to say here too.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
-                }
+        } else {
+            // need to name a setpoint node if using a setpoint type control mode
+            if ((state.dataPlantHXFluidToFluid->FluidHX(CompLoop).controlMode == ControlType::HeatingSetPointModulated) ||
```
Use `BITF` here.

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -191,336 +205,566 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
-            } else {
-                state.dataPlantHXFluidToFluid->FluidHX(CompLoop).SizingFactor = 1.0;
-            }
+        if (!lNumericFieldBlanks(5)) {
```
Conditional assignment (i.e., `x = y ? z : w`).

amirroth commented on src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:
```diff
@@ -531,16 +775,31 @@ void GetFluidHeatExchangerInput(EnergyPlusData &state)
+    std::string loopSupplySide;
+    std::string loopDemandSide;
+
+    if (this->Type == DataPlant::PlantEquipmentType::FluidToFluidPlantHtExchg) {
```
These can be `static std::array<std::string_view>` lookups.

amirroth commented on src/EnergyPlus/PlantLoadProfile.cc:
```diff
@@ -136,6 +138,7 @@ void PlantProfileData::simulate(EnergyPlusData &state,
     // Flow is requested and the actual available flow is set.  The outlet temperature is calculated.
 
     // Using/Aliasing
+    using FluidProperties::GetSatTemperatureRefrig;
```
I thought we are not doing this anymore.

amirroth commented on src/EnergyPlus/PlantLoadProfile.cc:
```diff
@@ -353,24 +414,55 @@ void GetPlantProfileInput(EnergyPlusData &state)
+
+            if (state.dataPlantLoadProfile->PlantProfile(ProfileNum).FluidType == PlantLoopFluidType::Water) {
+                state.dataPlantLoadProfile->PlantProfile(ProfileNum).InletNode =
+                    GetOnlySingleNode(state,
```
Why do we not get both `Inlet` and `Outlet` nodes in a single function call? This is a serious question. Maybe not for you, @dareumnam, but in general.

amirroth commented on src/EnergyPlus/PluginManager.cc:
```diff
@@ -263,6 +263,8 @@ void PluginManager::setupOutputVariables([[maybe_unused]] EnergyPlusData &state)
                     sResourceType = "DistrictCooling";
                 } else if (resourceType == "DISTRICTHEATING") {
                     sResourceType = "DistrictHeating";
+                } else if (resourceType == "DISTRICTHEATINGSTEAM") {
```
These functions, man.

amirroth commented on src/EnergyPlus/PollutionModule.cc:
```diff
@@ -2081,6 +2082,7 @@ void GetPollutionFactorInput(EnergyPlusData &state)
     FuelType.DieselFacilityIndex = GetMeterIndex(state, "Diesel:Facility");
     FuelType.PurchCoolFacilityIndex = GetMeterIndex(state, "DistrictCooling:Facility");
     FuelType.PurchHeatFacilityIndex = GetMeterIndex(state, "DistrictHeating:Facility");
+    FuelType.PurchHeatSteamFacilityIndex = GetMeterIndex(state, "DistrictHeatingSteam:Facility");
```
There is a better way here and we will implement it another day. 📌

amirroth commented on src/EnergyPlus/PollutionModule.cc:
```diff
@@ -6488,6 +6493,20 @@ void GetFuelFactorInfo(EnergyPlusData &state,
                 fuelSourceFactor = 1.084 / Pollution.PurchHeatEffic;
             }
 
+        } else if (SELECT_CASE_var == "DistrictHeatingSteam") {
```
Is `SELECT_CASE_var` on a `string` necessary? Don't we have this as an `enum` somewhere?

amirroth commented on src/EnergyPlus/WaterThermalTanks.hh:
```diff
@@ -232,7 +233,8 @@ namespace WaterThermalTanks {
                                                                                       "OtherFuel1",
                                                                                       "OtherFuel2",
-                                                                                      "DistrictHeating"};
+                                                                                      "DistrictHeating",
```
To the `.cc` file.

dareumnam committed:
DistrictHeating:Water and DistrictHeating:Steam 3

dareumnam committed:
Revert "DistrictHeating:Water and DistrictHeating:Steam 3"

This reverts commit e2c003d5515060cf11d693df6e6acf7169afa2d6.

dareumnam committed:
Revert "DistrictHeating:Water and DistrictHeating:Steam 2"

This reverts commit aa6f77fff6d22afaa9681d4998335ce09cc7857e.

dareumnam committed:
Revert "DistrictHeating:Water and DistrictHeating:Steam 1"

This reverts commit 3565765891eda3bab2deafa1c9adfedef704e8a7.

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
cleanup

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

nrel-bot-3 commented:
@dareumnam @myoldmopar it has been 28 days since this pull request was last updated.

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
separate DistrictHeating water and steam in idd

dareumnam committed:
Update NFP

dareumnam committed:
design doc

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
change DistrictHeating to DistrictHeatingWater and Add DistrictHeatingSteam 1

dareumnam committed:
change DistrictHeating to DistrictHeatingWater and Add DistrictHeatingSteam 2 ~ outputProcessor.cc

dareumnam committed:
change DistrictHeating to DistrictHeatingWater and Add DistrictHeatingSteam 2 ~ outputReportTabular.cc & PollutionModule.cc & hh

dareumnam committed:
change DistrictHeating to DistrictHeatingWater and Add DistrictHeatingSteam 4 ~ RuntimeLanguageProcessors.cc

dareumnam committed:
change DistrictHeating to DistrictHeatingWater and Add DistrictHeatingSteam 5 ~ WaterThermalTanks.cc

dareumnam committed:
Change DistrictHeating to DistrictHeatingWater in testfiles

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Change DistrictHeating to DistrictHeatingWater in /dataset and /performance_tests

dareumnam committed:
DistrictHeating to DistrictHeatingWater in unittests, idd, etc

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
fix custon check error

dareumnam committed:
clang format

dareumnam committed:
clang format

dareumnam committed:
fix unit tests and PlantLoadProfile 1

dareumnam committed:
Update Plantloadprofile 2

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -104732,7 +104865,7 @@ Output:EnvironmentalImpactFactors,
    \memo Used to help convert district and ideal energy use to a fuel type and provide total carbon equivalent with coefficients
    \memo Also used in Source=>Site conversions.
-  N1, \field District Heating Efficiency
+  N1, \field District Heating Water Efficiency
```
Should `N3, \field Steam Conversion Efficiency` change to `District Heating Steam Conversion Efficiency`?

dareumnam replied:
@mjwitte Thank you for catching this. I'll fix it.

dareumnam committed:
address comment

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
merge conflicts

dareumnam committed:
update testing files

EnergyArchmage commented on design/FY2021/NFP-ThreeNewSteamSystems.md:
```diff
@@ -0,0 +1,170 @@
+        \units m3/s
+        \ip-units gal/min
+        \autosizable
+   A5 , \field Water Inlet Node Name
```
I would expect that the steam connection to HX would be on the demand side of the steam loop.  And the water connection to HX would be on the supply side of a water loop.  The idea is that steam provides a source of heating for a water loop.

dareumnam replied:
@EnergyArchmage  Thanks for catching this. I did make the steam connection to HX is on the demand side of the steam loop, and the water connection to HX is on the supply side of a water loop. What I missed was fixing this \note in IDD and the design document, since I copied these lines from `HeatExchanger:FluidToFluid`. Thanks again for catching this. I'll fix this and I'll update the design doc by tonight.

dareumnam committed:
design doc

dareumnam committed:
Add fig 1

dareumnam committed:
delete backup file

dareumnam committed:
design doc

dareumnam committed:
design doc

dareumnam committed:
design doc

dareumnam committed:
design doc

EnergyArchmage commented on idd/Energy+.idd.in:
```diff
@@ -70888,10 +70888,26 @@ LoadProfile:Plant,
        \type object-list
        \object-list ScheduleNames
+  A6 , \field Plant Loop Fluid Type
+       \required-field
```
If these added inputs are optional for steam, then I doubt this should be a required field.

EnergyArchmage commented on src/EnergyPlus/PlantLoadProfile.cc:
```diff
@@ -258,6 +326,7 @@ void PlantProfileData::UpdatePlantProfile(EnergyPlusData &state) const
     // Set outlet node variables that are possibly changed
     state.dataLoopNodes->Node(this->OutletNode).Temp = this->OutletTemp;
+    state.dataLoopNodes->Node(this->OutletNode).MassFlowRate = this->MassFlowRate;
 }
```
Component models should not be setting plant node mass flow rates directly as is done here.  You already use central routine SetComponentFlowRate and it sets the node flow rates.  There could be other limitations outside of the component model that are affecting flow.  Managing that larger system gets unwieldly if component models do this sort of thing.

EnergyArchmage commented on src/EnergyPlus/OutsideEnergySources.cc:
```diff
@@ -427,19 +496,55 @@ void OutsideEnergySourceSpecs::calculate(EnergyPlusData &state, bool runFlag, Re
+            this->OutletTemp = SatTemp;
+            this->OutletSteamQuality = 1.0;
+            MyLoad = this->MassFlowRate * LatentHeatSteam;
+            state.dataLoopNodes->Node(this->OutletNodeNum).MassFlowRate = this->MassFlowRate;
```
also here, don't write to the node's massflow rate, just rely on SetComponentFlowRate above.

dareumnam committed:
Design doc

dareumnam committed:
add colon to IDD and input parts of districtheatingwater and districtheatingsteam

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
fix CI testing error & add two inputs to STWHX 1

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
fix CI testing errors

dareumnam committed:
update STWHX

dareumnam committed:
update Plp & districtheatingsteam

dareumnam committed:
update STWHX 2

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
update STWHX testing file

dareumnam committed:
Update PLP steam

dareumnam committed:
update testing file

dareumnam committed:
remove backup file

dareumnam committed:
fix custom check error, clang format, and remove one testing file that does not need

dareumnam commented:
Testing is still in progress and it seems like there is some issue with controlling the steam to water heat exchanger, but I think this is getting very close, I'd like to forge ahead so that we can get this done in a few weeks and hopefully this can get merged soon after this release.
the reaming work is
	• To resolve the control issue of steam to water heat exchanger
	• To resolve a few testings that fail in CI testing
	• To wrap up the transition rule for meter names/report variables CSV file
	• Documentation

dareumnam committed:
Pull develop and merge conflicts

dareumnam committed:
Add a testfile for STWHX

dareumnam committed:
update testfiles CMake list

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
fix merge conflicts

nrel-bot-2c commented:
@dareumnam @myoldmopar it has been 28 days since this pull request was last updated.

dareumnam committed:
pull develop and merge conflict

dareumnam committed:
fix integration error

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

nrel-bot-3 commented:
@dareumnam @myoldmopar it has been 28 days since this pull request was last updated.

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
remove =:

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
fix effectiveness check value

dareumnam committed:
pull develop and merge conflicts

mjwitte commented:
@dareumnam There are some conflicts here. Is this still targeted for v23.1 IOFreeze?

dareumnam commented:
@mjwitte This should be targeted for v23.2 IOFreeze. I'll remove the milestone.

nrel-bot-2 commented:
@dareumnam @myoldmopar it has been 28 days since this pull request was last updated.

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
fix errors

dareumnam committed:
Pull develop and merge conflicts

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
fix minor debug error

nrel-bot-2 commented:
@dareumnam it has been 7 days since this pull request was last updated.

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
fix error

dareumnam committed:
Merge conflicts, update stwhx testing file

dareumnam committed:
Merge branch 'AddThreeSteamModulesWithNTUMethod' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
clang format

dareumnam committed:
update functions

dareumnam committed:
update init func

dareumnam committed:
pull develop and merge conflicts

nrel-bot-2 commented:
@dareumnam @myoldmopar it has been 28 days since this pull request was last updated.

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
remove repeated lines

dareumnam committed:
Update DHS testfile

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
fix errors

dareumnam committed:
update districtheatingsteam

dareumnam committed:
update DistrictHeatingSteam & testfile

dareumnam committed:
Pull develop, merge conflicts

dareumnam committed:
fix errors, clang format

dareumnam committed:
clang format, update testfiles\CMakeLists.txt

dareumnam committed:
fix custom check error

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
update DHS, PLP Steam, testfiles, idd

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
update testfiles, add a comment

dareumnam committed:
remove indentations in the testfile

dareumnam committed:
fix warnings

dareumnam committed:
carve STWHX out

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
clang format

dareumnam committed:
fix error

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam commented:
1. New testfile: `SteamSystemAutoSize_DistrictHeatingSteam.idf`
 - Replace `Boiler:Steam` with `DistrictHeating:Steam` from `SteamSystemAutoSize.idf`
 - Every output variable shows a good match to a result from the original test file.
 
![image](https://github.com/NREL/EnergyPlus/assets/51240483/24d540b5-fcc5-4a54-bf26-fed080bb677a)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/253f28ef-5789-4f0a-aa8b-85cc721a258c)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/1e7600b4-63fe-4e79-b5fb-f0935a4e5fca)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/0e2db130-9a27-449f-bd39-839e10f237e8)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/edfbf1ca-7981-4628-b92d-a5e9573e56cc)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/ee10a341-204c-49e4-a59e-b641a3eeac13)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/6c8ebde9-9637-4a47-bfde-d86c91fa0945)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/ddcf8860-ef16-43c9-8436-86dbb74f3ce6)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/ffe1e8c3-af5d-44ac-8eca-2bbbcd1cad3a)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/d1909b0c-64fa-4113-8e8c-8f888bc42b4b)

dareumnam commented:
2. New testfile: `PlantLoadProfileSteam.idf`
 - The testfile is a plant-only simulation using the PLANT LOAD PROFILE object for a steam loop with `Boiler:Steam`. There are no simulated zones.
 - The load profile heat transfer rate matches the load schedule input. The steam boiler heating rate, the plant loop supply side temperature/mass flow rate, the demand side temperature, and the mass flow rate all appear reasonable.

![image](https://github.com/NREL/EnergyPlus/assets/51240483/ba430e84-d56a-469d-85f2-68fda6a12a56)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/6f3789b4-f2d6-418b-bc17-cba0a2f90dbc)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/7b2f37d9-663f-42bc-b7fe-ea67bd5a76d3)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/3ebc7e9e-0de4-4217-b509-eb8283afaf65)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/7c24ccef-3b50-4a7d-923b-4e12265850e8)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/375c01c4-800e-45b4-ab90-f8ed7b0e8ba4)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/efcdafae-59b7-4141-b746-ec42e5f5c729)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/a7edad1a-d374-44a8-ade8-cfb58893d009)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/59b7b2f4-3eb2-4ace-ba79-f6ddaf667ff7)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/e45605bf-9616-4541-801a-24fa09229b40)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/8ac16358-7676-4f38-94a5-9cb6857dbb68)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/f4cabde2-eac5-4592-b952-f2d49c1bfe46)
(↑ The higher value is 100, and the lower value is 99.9999999999999)
![image](https://github.com/NREL/EnergyPlus/assets/51240483/4e973e39-58ec-406e-8594-bfd01341204f)

rraustad commented:
Very well documented example files. Nice job!

dareumnam commented:
(Currently, I'm working on the failed tests.)

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -71820,10 +71820,26 @@ LoadProfile:Plant,
+       \note optional, in case of steam loop
+       \units C
+       \minimum 10.0
+       \default 20.0
```
Some idd examples of use of the word optional in note:

    \note optional                 <- I don't like this one since it does not give any other guidance
    \note optional (only required for thermal comfort runs)            <- this is better

Better use of note to guide user (these are more typical of fields that connect to other fields):

    \note Used only when Optical Data Type = Spectral
    \note This field is used only when Outdoor Air Method=DetailedSpecification. <- I like this one

So your notes could say:

    \note This field is used only when Plant Loop Fluid Type=Steam. 

It's really up to you since there are no hard rules for notes but it's better to maintain consistency throughout the idd.

dareumnam replied:
Thank you for your insight. I'll update the \note to maintain consistency throughout the idd.

dareumnam committed:
fix failed tests & address comment

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

rraustad commented on src/EnergyPlus/OutsideEnergySources.cc:
```diff
@@ -428,21 +472,48 @@ void OutsideEnergySourceSpecs::calculate(EnergyPlusData &state, bool runFlag, Re
+            this->MassFlowRate = state.dataLoopNodes->Node(this->OutletNodeNum).MassFlowRate;
+            this->OutletTemp = state.dataLoopNodes->Node(state.dataPlnt->PlantLoop(this->plantLoc.loopNum).TempSetPointNodeNum).TempSetPoint;
+            // Like the assumption in Boiler:Steam, assume that it can meet the steam loop setpoint
+            this->OutletSteamQuality = 1.0;
```
Is this right? Quality = 1? The mass flow rate is calculated to bring the steam all the way to EnthSteamOutWet, which would be a quality of 0. See line 505 above, and 508 and 509. And at line 520, Quality would be 1.

dareumnam replied:
I believe that the outlet steam quality of the district heating steam is 1. (And the inlet steam quality of the district heating steam is 0.) According to EnergyPlus' assumptions for the steam system, the steam loop in EnergyPlus is designed to operate under saturated conditions, with no superheated vapor. Also, a `Boiler:Steam` generates steam with a quality of 1 each time. Similarly, the `DistrictHeatingSteam` component is expected to provide steam with a quality of 1 to the demand side of the steam loop.  `EnthSteamInDry`, `EnthSteamOutWet`, and `LatentHeatSteam` are interim variables that are used to calculate the `SteamMdot`.

rraustad replied:
Yes, the quality of the steam leaving the district heating plant = 1. But this calculation is not the quality of the steam from the district heating plant. It's the quality of the steam after it is used to meet a load for outside energy sources, specifically MyLoad.

If I look at the call for EnthSteamOutWet I see Quality = 0 in argument 4. While Quality = 1 for EnthSteamInDry. That makes sense, right? Then:

    LatentHeatSteam = EnthSteamInDry - EnthSteamOutWet;, and
    SteamMdot = MyLoad / LatentHeatSteam;

So the model is setting the steam mass flow rate to pull all the latent heat out of the fluid. If that happens, and it will if Mdot = SteamMdot, which may not be the case after the call to SetComponentFlowRate, then all the latent heat was removed. That means the fluid quality has changed. It may not get to 0 (if the plant limits the steam flow rate), but it is certainly not 1. And quality will usually be 0 at this point (if flow rate is not limited).

    Real64 EnthSteamOutWet = FluidProperties::GetSatEnthalpyRefrig(state,
                                                                           state.dataPlnt->PlantLoop(this->plantLoc.loopNum).FluidName,
                                                                           this->InletTemp,
                                                                           0.0,
                                                                           state.dataPlnt->PlantLoop(this->plantLoc.loopNum).FluidIndex,
                                                                           RoutineName);

dareumnam replied:
I got your point. Thanks for the explanation. I made the component member variable `OutletSteamQuality` and regarded it as the quality of steam leaving the district heating steam plant. I'll fix the code as per your insight. Or, I could remove this member variable since it's pretty much a useless variable.

rraustad replied:
I think just leave the variable for now, just make sure it reports the correct value. Quality = 1 for steam, quaulity = 0 for water at 100 C. If the component is on quality out = 0, if component is off quality out = 1,

rraustad replied:
I don't even know what quality means in the context of steam in E+. I mean it's steam, 0 means it's hot water. What does 0.5 mean? in an open system. If it means the fluid is half way to being water then I am totally lost. It;s an enthalpy question.

dareumnam replied:
The assumption that it cannot allow superheated vapor makes everything simple but a little awkward. It would be much better if EnergyPlus's steam system could ultimately be upgraded to an enthalpy-based system, enabling the use of superheated vapor (someday, if funding allows…).

rraustad commented on src/EnergyPlus/PlantLoadProfile.cc:
```diff
@@ -126,38 +123,79 @@ void PlantProfileData::simulate(EnergyPlusData &state,
+            this->OutletTemp = SatTemp - this->LoopSubcoolReturn;
+        } else {
+            this->Power = 0.0;
+        }
```
There is no OutletSteamQuality here? as there is in SteamCoils and OutsideEnergySources. Actually it's pretty much a useless variable since it's either 1 or 0 but it does represent that the fluid either has potential energy or does not.

dareumnam replied:
Thank you for catching this. I'll set OutletSteamQuality here.

dareumnam replied:
Or we might remove `OutletSteamQuality` here and from `DistrictHeating:Steam`. In the case of `Boiler:Steam`, it does not have component member variables about steam quality, and it directly updates the steam quality data of LoopNode like below.
`state.dataLoopNodes->Node(BoilerOutletNode).Quality = 1.0; `

rraustad commented:
@Myoldmopar I made another pass on this with only 2 comments. Also, is there no unit test for Purchased*? Not saying there is more work to do here, just noting I don't see a unit test specific to these objects. I do see that these objects are used in many unit tests which may be good enough for coverage.

dareumnam commented:
@rraustad Thanks for the comment. "✅Adding a few unit tests for new objects" is on my to-do list for this PR and I am hoping I can get this done before I/O freeze.🏃‍♀️

rraustad commented on src/EnergyPlus/PlantLoadProfile.hh:
```diff
@@ -65,27 +65,38 @@ struct EnergyPlusData;
+        Water,
+        Steam,
+        Num
+    };
```
Is glycol not an option? or is that just another form of water?

dareumnam replied:
I think it is another form of water. I set the enum `Water` as per the NodeFluidType of PlantLoadProfile's inlet/outlet node [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/PlantLoadProfile.cc#L361).

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into AddThreeSteamModulesWithNTUMethod

dareumnam committed:
Add loop limits on steam mass flow rate result to keep in check

dareumnam committed:
address comments

dareumnam committed:
address comments

dareumnam commented:
TO-DO list for the PR:
 
 - [ ] Address the comments
 - [ ] As far as Meter name, `DistrictHeatingSteam` should be reverted to `steam` (due to steam as `Generator Heat Source Type` of `Chiller:Absorption` and `Chiller:Absorption:Indirect`, where the generator inlet and outlet nodes are connected to a steam loop with `Boiler:Steam`)
 - [ ] Transition: input rules:
   - [ ] Object name change (`DistrictHeating` to `DistrictHeating:Water`)
   - [ ] Object change (`LoadProfile:Plant`, insert three new inputs at the end)
   - [ ] Output changes: `DistrictHeating` to `DistrictHeatingWater` as ResourceType/Meter key
 - [ ] Make a few unit tests

dareumnam committed:
Revert Meter Key DistrictHeatingSteam to Steam

dareumnam committed:
fix the failed unit tests

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -106442,7 +106483,7 @@ EnvironmentalImpactFactors,
       \minimum> 0.0
       \default 3.0
-  N3, \field Steam Conversion Efficiency
+  N3, \field District Heating Steam Conversion Efficiency
```
Remember to mention field name changes in the rules description - important for epJSON files.

mjwitte commented on src/EnergyPlus/EconomicTariff.cc:
```diff
@@ -3864,51 +3864,63 @@ void LEEDtariffReporting(EnergyPlusData &state)
-    DemandWindow distHeatDemWindowUnits;
+    DemandWindow distHeatWaterDemWindowUnits;
+    DemandWindow distHeatSteamDemWindowUnits;
     DemandWindow othrDemWindowUnits;
 
     auto const &tariff = state.dataEconTariff->tariff;
 
     if (state.dataEconTariff->numTariff > 0) {
         int distCoolFacilMeter = GetMeterIndex(state, "DISTRICTCOOLING:FACILITY");
-        int distHeatFacilMeter = GetMeterIndex(state, "DISTRICTHEATING:FACILITY");
+        int distHeatWaterFacilMeter = GetMeterIndex(state, "DISTRICTHEATINGWATER:FACILITY");
+        int distHeatSteamFacilMeter = GetMeterIndex(state, "DISTRICTHEATINGSTEAM:FACILITY");
```
For transition, do we assume that all existing districtheating is transitioned to DistrictHeatingWater?

Myoldmopar replied:
This looks like a question, but @dareumnam gave a :+1: so I think that's a :+1:

dareumnam replied:
Yes for now, all existing `Districtheating` is transitioned to `DistrictHeatingWater` since existing `Districtheating` is about water loop, but if you have any concerns/suggestions, please let me know.

mjwitte commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -3997,7 +4002,7 @@ namespace OutputProcessor {
             state.dataOutputProcessor->indexGroupKey = 101;
-        } else if (has(meterName, "DistricHeating:Facility")) {
+        } else if (has(meterName, "DistricHeatingWater:Facility")) {
             state.dataOutputProcessor->indexGroupKey = 102;
```
Hmm, this is interesting. I don't know how these `indexGroupKey` values were used in the past, but it appears these are never used. I'm 99% certain all use of this can be removed. Something to note for a followup PR.

Myoldmopar replied:
Always good to identify dead code :+1:

mjwitte commented on src/EnergyPlus/OutputReportPredefined.cc:
```diff
@@ -1192,8 +1192,10 @@ namespace OutputReportPredefined {
         s->pdchLeedPerfDisClDem = newPreDefColumn(state, s->pdstLeedPerf, "District Cooling Demand [W]");
-        s->pdchLeedPerfDisHtEneUse = newPreDefColumn(state, s->pdstLeedPerf, "District Heating Use [GJ]");
-        s->pdchLeedPerfDisHtDem = newPreDefColumn(state, s->pdstLeedPerf, "District Heating Demand [W]");
+        s->pdchLeedPerfDisHtWtrEneUse = newPreDefColumn(state, s->pdstLeedPerf, "District Heating Water Use [GJ]");
+        s->pdchLeedPerfDisHtWtrDem = newPreDefColumn(state, s->pdstLeedPerf, "District Heating Water Demand [W]");
+        s->pdchLeedPerfDisHtStEneUse = newPreDefColumn(state, s->pdstLeedPerf, "District Heating Steam Use [GJ]");
+        s->pdchLeedPerfDisHtStDem = newPreDefColumn(state, s->pdstLeedPerf, "District Heating Steam Demand [W]");
```
These table output column changes (and any others) need to be noted in OutputChanges23-1-0-to-23-2-0.md

mjwitte reviewed (commented):
@dareumnam
There are some err diffs like this:
```
+   ** Warning ** The following Report Variables were requested but not generated -- check.rdd file
+   **   ~~~   ** Either the IDF did not contain these elements, the variable name is misspelled,
+   **   ~~~   ** or the requested variable is an advanced output which requires Output : Diagnostics, DisplayAdvancedReportVariables;
+   ************* Key=*, VarName=DISTRICT HEATING HOT WATER RATE, Frequency=Hourly
+   ************* Key=*, VarName=DISTRICT HEATING HOT WATER ENERGY, Frequency=Hourly

+   ** Warning ** The following Report Variables were requested but not generated -- check.rdd file
+   **   ~~~   ** Either the IDF did not contain these elements, the variable name is misspelled,
+   **   ~~~   ** or the requested variable is an advanced output which requires Output : Diagnostics, DisplayAdvancedReportVariables;
+   ************* Key=*, VarName=DISTRICT COOLING CHILLED WATER RATE, Frequency=Hourly
+   ************* Key=*, VarName=DISTRICT COOLING CHILLED WATER ENERGY, Frequency=Hourly
```
And one unit test failure.

Have you written up transition rules? @Myoldmopar or I can write the transition code if you write the rules in `\src\Transition\InputRulesFiles\Rules23-1-0-to-23-2-0.md`.

This also needs a list of meter name changes in `\src\Transition\SupportFiles\Report Variables 23-1-0 to 23-2-0.csv`

dareumnam commented:
@mjwitte Thank you for the reviews. Yes, I fixed failed unit test locally, and I'm also currently working on Transition rules including `Rules23-1-0-to-23-2-0.md`, `CreateNewIDFUsingRulesV23_2_0.f90`,  `OutputChanges23-1-0-to-23-2-0.md`, and `Report Variables 23-1-0 to 23-2-0.csv` locally. I'll push these commits as soon as possible. I'll also summarize the test diffs. 🏃‍♀️🏃‍♀️

dareumnam committed:
Transition: update CreateNewIDFUsingRulesV23_2_0 and Rules23-1-0-to-23-2-0

dareumnam commented:
TO-DO list for the PR:
 
 - [x] Address the comments
 - [x] As far as Meter name, `DistrictHeatingSteam` should be reverted to `steam` (due to steam as `Generator Heat Source Type` of `Chiller:Absorption` and `Chiller:Absorption:Indirect`, where the generator inlet and outlet nodes are connected to a steam loop with `Boiler:Steam`)
 - [x] Transition
   - [x] Input rules: Object name change (`DistrictHeating` to `DistrictHeating:Water`)
   - [x] Input rules: Object change (`LoadProfile:Plant`, insert three new inputs at the end)
   - [x] Input rules: Choice-field of Resource Type/Fuel Type (`DistrictHeating` to `DistrictHeating:Water`, `Steam` to `DistrictHeatingSteam`) in `OtherEquipment`, `Exterior:FuelEquipment`, `ZoneHVAC:HybridUnitaryHVAC`, `WaterHeater:Mixed`, `WaterHeater:Stratified`, `EnergyManagementSystem:MeteredOutputVariable`, `Meter:Custom`, `Meter:CustomDecrement`, and `PythonPlugin:OutputVariable`
   - [x] Output changes: `DistrictHeating` to `DistrictHeatingWater` as ResourceType/Meter key
   - [x] Update output change report variables CSV
 - [x] Summarize test diffs: Mostly test diffs are table heading, and table size mismatching (As for Demand End Use Components Summary Report, the current develop shows either DistrictHeating or Steam, while this branch shows both DistrictHeatingWater and DistrictHeatingSteam.)

Follow-up PR after I/O freeze:
 - [ ] Make a few unit tests
 - [ ] Documentation: Engineering Reference and I/O Reference

Myoldmopar commented on src/EnergyPlus/Plant/PlantManager.cc:
```diff
@@ -1206,6 +1206,11 @@ void GetPlantInput(EnergyPlusData &state)
+                    case PlantEquipmentType::PurchSteam: {
+                        this_comp.compPtr =
+                            OutsideEnergySources::OutsideEnergySourceSpecs::factory(state, PlantEquipmentType::PurchSteam, CompNames(CompNum));
+                        break;
```
Does this object need to have the `FlowPriority` and `HowLoadServed` variables set to get flow working properly?  I guess maybe not, since the one above it doesn't have it.

Myoldmopar commented on src/Transition/CreateNewIDFUsingRulesV23_2_0.f90:
```diff
@@ -674,9 +674,35 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
+                nodiff=.false.
+                ObjectName='DistrictHeating:Water'
+                CALL GetNewObjectDefInIDD(ObjectName,NwNumArgs,NwAorN,NwReqFld,NwObjMinFlds,NwFldNames,NwFldDefaults,NwFldUnits)
+                OutArgs(1:CurArgs)=InArgs(1:CurArgs)
```
Yep this is what I would expect for an object name change

Myoldmopar commented on src/Transition/CreateNewIDFUsingRulesV23_2_0.f90:
```diff
@@ -674,9 +674,35 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
+                 END IF
+                 IF (OutArgs(5) == "STEAM" .OR. OutArgs(5) == "Steam" .OR. OutArgs(5) == "steam") THEN
+                     OutArgs(5) = "DistrictHeatingSteam"
+                 END IF
```
This object has min-fields 7 so this field should always be there.  However, the upper/lower case needs work here.  I can't recall the Fortran string comparison with trailing spaces.  OutArgs(5) is actually like 100 characters long filled with blanks.  So at a minimum I think we need to TRIM the OutArgs(5) to be sure we are comparing just the non-blank characters.  But more than that, what if the user enters DISTRICTHeating?  That's not a case here.  We should do something different -- I think Transition has SameString which will handle all these things nicely.  

```
IF (SameString(OutArgs(5), "DistrictHeating") THEN
```

should work I think.

dareumnam commented:
@mjwitte 's comment in `ChillerAbsorption.cc` solved all these questions. I need to revert the commit though :)
~~I would appreciate your insights on this matter.~~

~~- As far as Meter Type and Resource Type, I replaced `DistrictHeating` with `DistrictHeatingWater` and `Steam` with `DistrictHeatingSteam`. This was the conclusion that everyone agreed upon during the EnergyPlus Technicalities Call a year ago.~~

~~- `Chiller:Absorption` and `Chiller:Absorption:Indirect` have an input for `Generator Heat Source Type`, as shown below. As the note indicates, they can be connected to a steam loop or not. In cases where the chiller object is not connected to a steam loop (like in `AbsorptionChiller.idf`), `DistrictHeatingSteam` is the correct resource type. However, if the chiller object is connected to a steam loop (as seen in `IndirectAbsorptionChiller.idf`), `DistrictHeatingSteam` is not the appropriate resource type.~~

~~- So I reverted `DistrictHeatingSteam` to `Steam` only for the output sections [here](https://github.com/NREL/EnergyPlus/pull/9260/commits/3c8b232e9abe6239d99a63646c18fd9aa0169412#diff-bd9d5281a7fe99bacbcd3d7112772f7804d8e091f10a56a0954a7639c24384fcR1451-R1463). The input sections and the resource type enumeration remain as `DistrictHeatingSteam`. However, I am uncertain whether this approach is acceptable or not. I would appreciate your insights on this matter. (For example, in `Meter:Custom`, the Resource Type input is presented `DistrictHeatingSteam` as a choice-field, but the Meter as output is going to display `Steam`.)~~

Myoldmopar commented on src/Transition/InputRulesFiles/Rules23-1-0-to-23-2-0.md:
```diff
@@ -491,3 +491,119 @@ See [PR#9925](https://github.com/NREL/EnergyPlus/pull/9925)
+Field 1 has a name change from "District Heating Efficiency" to "District Heating Water Efficiency".
+Field 3 has a name change from "Steam Conversion Efficiency" to "District Heating Steam Conversion Efficiency".
+
+See [PR#9260](https://github.com/NREL/EnergyPlus/pull/9260
```
I think these rules are sufficient, thanks.

Myoldmopar reviewed (commented):
I've got some transition updates locally that I need to finish testing, but otherwise this looks good.

mjwitte commented on src/EnergyPlus/ChillerAbsorption.cc:
```diff
@@ -651,7 +651,7 @@ void BLASTAbsorberSpecs::setupOutputVars(EnergyPlusData &state)
                                 this->Name,
                                 {},
-                                "DistrictHeatingSteam",
+                                fluidNameSteam,
```
@dareumnam If I'm following correctly,I think this should be "DistrictHeatingSteam". This for for SetupOutputVariable is only used it the chiller is not connected to a plant loop, so it is consuming "DistrictHeatingSteam" directly. If the chiller is conncted to a plant loop, then the SetupOutputVariable call above is used which sets the resource type to PLANTLOOPHEATINGDEMAND.

mjwitte commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -1448,14 +1448,19 @@ namespace OutputProcessor {
         std::string const meterType = UtilityRoutines::makeUPPER(UserInputResourceType);
 
         int eMeterResource = getEnumValue(Constant::eResourceNamesUC, meterType);
+        if (meterType == "STEAM") {
```
This shouldn't be necessary.

dareumnam committed:
Update OutputRules, Output report vars CSV, Revert commits about steam meter names, etc

dareumnam committed:
Add design doc, remove FlowPriority and HowLoadServed for DistrictHeatingSteam

dareumnam commented:
I just found the same variable `this->EnergyRate` was set as two different output variables in the existing DistrictCooling/Heating objects [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/OutsideEnergySources.cc#L497-L511). Should one of these be removed? @mjwitte @Myoldmopar @rraustad

rraustad commented:
I see EnergyTransfer just above that for cooling or heating. I think @mjwitte or @Myoldmopar need to answer this one.

Myoldmopar commented:
The documentation shows that this variable is duplicated intentionally for now, leave it

![image](https://github.com/NREL/EnergyPlus/assets/849824/10273cc7-7be3-4d5b-8dcd-e7fc7423006e)

mjwitte commented:
> I just found the same variable `this->EnergyRate` was set as two different output variables in the existing DistrictCooling/Heating objects [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/OutsideEnergySources.cc#L497-L511). Should one of these be removed? @mjwitte @Myoldmopar @rraustad

In develop, you get these outputs:
HVAC,Average,District Cooling Chilled Water Rate [W]
HVAC,Average,District Cooling Rate [W]
HVAC,Average,District Heating Hot Water Rate [W]
HVAC,Average,District Heating Rate [W]

With this branch, we get just these in the rdd, but we get two instances in the csv output, because the same name is getting set up twice.
HVAC,Average,District Cooling Water Rate [W]
HVAC,Average,District Heating Water Rate [W]
HVAC,Average,District Heating Steam Rate [W]

So, we need to change something here.  Since they are the same value, we could just set it up once. Or we could think of a way to change the name slightly and keep both.

dareumnam committed:
Update Report Variable CSV

mjwitte commented:
@Myoldmopar @rraustad @dareumnam I propose we just keep one set, and delete the second SetupOutputVariable call.
HVAC,Average,District Cooling Water Rate [W]
HVAC,Average,District Heating Water Rate [W]
HVAC,Average,District Heating Steam Rate [W]

Docs and Report Variables 23-1-0 to 23-2-0.csv need to reflect these changed output variable names. In Report Variables 23-1-0 to 23-2-0.csv map both of the old variables to the same new variable, so two rows for heating and two rows for cooling.

dareumnam commented:
@mjwitte Thanks for your insight, it sounds great to me. I'll delete the second SetupOutputVariable call, and update Report Variables 23-1-0 to 23-2-0.csv and Docs after @Myoldmopar and @rraustad confirm this.

Myoldmopar commented:
Yeah that makes sense.  I initially thought we might just delete one of them in the report vars file, but it could lead to random removals.  Mapping them together could end up with a minor warning in some IDFs, but it's reasonable.

dareumnam committed:
Delete the second SetupOutputVariable call and Update Report Variable CSV

dareumnam committed:
Update Report Variable CSV and change updated output var names in testfiles

dareumnam committed:
update NFP

dareumnam committed:
fix CSV typos

mjwitte commented:
There are two err diffs that show some idf file changes are needed.
[ExteriorLightsAndEq](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2023-08/273aede840425898126ac7f34ce658436f17297f-f425e2afde5653c8af2107fb544c1c58c7a881f0/ExteriorLightsAndEq/x86_64-Linux-Ubuntu-22.04-gcc-11.4/eplusout.err.diff.html)
[RadLoTempCFloHeatCool_AddedAutosizing](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2023-08/273aede840425898126ac7f34ce658436f17297f-f425e2afde5653c8af2107fb544c1c58c7a881f0/RadLoTempCFloHeatCool_AddedAutosizing/x86_64-Linux-Ubuntu-22.04-gcc-11.4/eplusout.err.diff.html)

Myoldmopar commented:
Thanks @mjwitte.  I'm going to try to get the branch all fixed up including those issues.   @dareumnam don't do any more commits/pushes for now, but I'll let you know if I need anything from you after my testing/work is done.

dareumnam commented:
@mjwitte @Myoldmopar Thanks so much for the review and fix. Please let me know if there is anything further I can do.

Myoldmopar commented:
I am hopeful that I will have this ready for merge soon.  I realized this morning that I was running tests using a binary from a different build folder, which was causing some spurious issues, most importantly just wasting test time.  I am now seeing good results coming out and will report back soon.  Once I'm happy locally, I'll push up my transition changes and any other changes needed.

mjwitte committed:
Cleanup District sizing output

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into AddThreeSteamModulesWithNTUMethod

dareumnam commented:
There might be more comments that I missed a year ago. I'll address them in the follow-up PR after the I/O freeze.

mjwitte commented on src/EnergyPlus/OutsideEnergySources.cc:
```diff
@@ -386,12 +424,12 @@ void OutsideEnergySourceSpecs::size(EnergyPlusData &state)
         }
         if (!this->NomCapWasAutoSized && this->NomCap > 0.0 && state.dataPlnt->PlantFinalSizesOkayToReport) {
-            BaseSizer::reportSizerOutput(state, "District" + typeName, this->Name, "User-Specified Nominal Capacity [W]", this->NomCap);
+            BaseSizer::reportSizerOutput(state, format("{}", typeName), this->Name, "User-Specified Nominal Capacity [W]", this->NomCap);
```
@Myoldmopar I missed this one, can be simplified to `state, typeName, this->Name, ...` but it works as-is.

Myoldmopar commented:
Local testing is much, much better on the last run.  There is still a class of warnings coming out that I need to clean up, but it's much better now.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into AddThreeSteamModulesWithNTUMethod

mjwitte committed:
Cleanup District sizing output

Myoldmopar committed:
Updated transtiion rules and report variables file.  All test files pass after transition.

Myoldmopar commented:
New warnings are now gone, the only thing I had forgotten (it appears) was to transition the component setpoint operation object, which also includes the object type.  It caused one file to fail after transitioning them all.  I fixed that, retested and it worked.  I believe this is ready for merge now.  I will let Decent run tonight, and if clean, it will join the ruleset branch in the merge queue.

Myoldmopar commented:
Alright, CI says:
- Windows all tests pass, great
- Linux debug tests pass, great
- Linux regressions (confirmed locally):
  - Audit diffs are fine
  - Table diffs look related to a new defaulted field input
  - MDD/RDD diffs due to report variable name change
  - CSV - not diffs really, but couldn't match some column names due to variable name changes
  - BND changes due to object name change

I think this is sufficiently ready to go, merging.  Thanks @dareumnam for this effort.  This is a great step forward in being able to simulate steam systems.  Thanks @mjwitte and @rraustad for the reviews.

jmarrec commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -1750,8 +1750,8 @@ void GetInputOutputTableSummaryReports(EnergyPlusData &state)
-        ort->resourceTypeNames(4) = "DistrictHeating";
-        ort->resourceTypeNames(5) = "Steam";
+        ort->resourceTypeNames(4) = "DistrictHeatingWater";
+        ort->resourceTypeNames(5) = "DistirctHeatingSteam";
```
Typo!

Root cause of #10190 . That's a huge impact for a small typo, init?
